src/ar_tx_db.erl

```diff
@@ -22,33 +21,49 @@ start() ->
 
 %% @doc Put an Erlang term into the meta DB. Typically these are
 %% write-once values.
-put(Key, Val) ->
-    ets:insert(?MODULE, {Key, Val}),
-    timer:apply_after(1800*1000, ?MODULE, remove, [Key]).
+put_error_codes(TXID, ErrorCodes) ->
+	ets:insert(?MODULE, {TXID, ErrorCodes}),
+	{ok, _} = timer:apply_after(1800*1000, ?MODULE, remove, [TXID]),
+	ok.
+
 %% @doc Retreive a term from the meta db.
-get(Key) ->
-	case ets:lookup(?MODULE, Key) of
-		[{Key, Obj}] -> Obj;
+get_error_codes(TXID) ->
+	case ets:lookup(?MODULE, TXID) of
+		[{_, ErrorCodes}] -> {ok, ErrorCodes};
 		[] -> not_found
 	end.
 
-maybe_add(Key) ->
-    case ets:lookup(?MODULE, Key) of
-        [{Key, Value}] -> ok;
-        _ -> put(Key, ["unknown_error "])
-    end.
+%% @doc Writes an unknown error code if there are not already any error codes
+%% for this TX.
+ensure_error(TXID) ->
+	case ets:lookup(?MODULE, TXID) of
+		[_] -> ok;
+		[] -> put_error_codes(TXID, ["unknown_error"])
+	end.
 
-remove(Key) ->
-    ets:delete(?MODULE, Key).
+%% @doc Removes all error codes for this TX.
+clear_error_codes(TXID) ->
+	ets:delete(?MODULE, TXID).
 
 tx_db_test() ->
-  	ar_storage:clear(),
-	{Priv1, Pub1} = ar_wallet:new(),
-	[B0] = ar_weave:init([{ar_wallet:to_address(Pub1), ?AR(10000), <<>>}]),
-	OrphanedTX = ar_tx:new(Pub1, ?AR(1), ?AR(5000), <<>>),
-	TX = OrphanedTX#tx { owner = Pub1 , signature = <<"BAD">>},
-	SignedTX = ar_tx:sign(TX, Priv1, Pub1),
-	ar_tx:verify(TX, 8, B0#block.wallet_list),
+	ar_storage:clear(),
+	{_, Pub1} = ar_wallet:new(),
+	{Priv2, Pub2} = ar_wallet:new(),
+	[B0] = ar_weave:init([
+		{ar_wallet:to_address(Pub1), ?AR(10000), <<>>},
+		{ar_wallet:to_address(Pub2), ?AR(10000), <<>>}
+	]),
+	%% Test bad transaction
+	OrphanedTX1 = ar_tx:new(Pub1, ?AR(1), ?AR(5000), <<>>),
+	BadTX = OrphanedTX1#tx { owner = Pub1, signature = <<"BAD">> },
+	?assert(not ar_tx:verify(BadTX, 8, B0#block.wallet_list)),
 	timer:sleep(500),
-    ["tx_too_cheap ","tx_fields_too_large ","tag_field_illegally_specified ","last_tx_not_valid "] = get(TX#tx.id),
-    ar_tx:verify(SignedTX, 8, B0#block.wallet_list).
+	Expected = {ok, ["same_owner_as_target", "tx_id_not_valid", "tx_signature_not_valid"]},
+	?assertEqual(Expected, get_error_codes(BadTX#tx.id)),
+	%% Test good transaction
+	OrphanedTX2 = ar_tx:new(Pub1, ?AR(1), ?AR(5000), <<>>),
+	SignedTX = ar_tx:sign(OrphanedTX2, Priv2, Pub2),
+	?assert(ar_tx:verify(SignedTX, 8, B0#block.wallet_list)),
+	clear_error_codes(BadTX#tx.id),
+	clear_error_codes(SignedTX#tx.id),
```

Comment:
Makse sense to add an additional assertion to assert `clear_error_codes` actually cleans it up.

Reply:
`ets:delete/2` always return `true`, so there is nothing to test.

Reply:
You can call `get_error_codes` after `clear_error_codes` and observe the absence of some previously added entries.

Reply:
How about now?